graphql/src/main/java/com/scalar/db/graphql/datafetcher/TransactionInstrumentation.java

```diff
@@ -0,0 +1,145 @@
+import com.scalar.db.exception.transaction.TransactionException;
+import com.scalar.db.exception.transaction.UnknownTransactionStatusException;
+import com.scalar.db.graphql.schema.Constants;
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.GraphQLContext;
+import graphql.execution.AbortExecutionException;
+import graphql.execution.ExecutionContext;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecuteOperationParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.language.Argument;
+import graphql.language.BooleanValue;
+import graphql.language.Directive;
+import graphql.language.StringValue;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+public class TransactionInstrumentation extends SimpleInstrumentation {
+  @VisibleForTesting static final String EXTENSIONS_TRANSACTION_KEY = "transaction";
+  @VisibleForTesting static final String EXTENSIONS_TX_ID_KEY = "txId";
+
+  @VisibleForTesting final Map<String, DistributedTransaction> transactionMap;
+  private final DistributedTransactionManager transactionManager;
+
+  public TransactionInstrumentation(DistributedTransactionManager transactionManager) {
+    this.transactionManager = transactionManager;
+    this.transactionMap = new HashMap<>();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecuteOperation(
+      InstrumentationExecuteOperationParameters parameters) {
+    ExecutionContext executionContext = parameters.getExecutionContext();
+    GraphQLContext graphQLContext = executionContext.getGraphQLContext();
+
+    List<Directive> transactionDirectives =
+        executionContext
+            .getOperationDefinition()
+            .getDirectives(Constants.TRANSACTION_DIRECTIVE_NAME);
+
+    if (transactionDirectives.isEmpty()) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Directive directive = transactionDirectives.get(0); // @transaction is not repeatable
+    Argument txIdArg = directive.getArgument(Constants.TRANSACTION_DIRECTIVE_TX_ID_ARGUMENT_NAME);
+    Argument commitArg =
+        directive.getArgument(Constants.TRANSACTION_DIRECTIVE_COMMIT_ARGUMENT_NAME);
+    boolean isCommitTrue =
+        commitArg != null
+            && commitArg.getValue() != null
+            && ((BooleanValue) commitArg.getValue()).isValue();
+    if (txIdArg == null) {
+      // start a new transaction
+      try {
+        DistributedTransaction transaction = transactionManager.start();
+        transactionMap.put(transaction.getId(), transaction);
```

Comment:
When do we delete completed transaction objects committed or aborted from this `transactionMap`?

Also I think we need to handle the situation where a client is down while executing a transaction. In this case, the client doesn't call either commit or abort the transaction. So we need to remove transactions in `transactionMap` for a long time.
Maybe, we can use `ActiveExpiringMap` for that case:
https://github.com/scalar-labs/scalardb/blob/df1ba719bd55bb15ca7554762ec27ee90417f258/core/src/main/java/com/scalar/db/util/ActiveExpiringMap.java

Reply:
Yes, the transaction needs to be removed after a certain timeout if the client is down. `ActiveExpiringMap` sounds like a perfect option. Thank you for the great suggestion!
Also, I carelessly deleted a line that deletes transaction objects committed or aborted. I will update the code and tests.

Reply:
Now an `ActiveExpiringMap` is used to keep transactions. Please take a look when you have time.